zkml-jolt-core/src/jolt/fp_lookups/mod.rs

```diff
@@ -0,0 +1,1239 @@
+impl FpLookupInstance {
+    /// Create a new FpLookupInstance from an instruction
+    pub fn new(
+        instr: &onnx_tracer::trace_types::AtlasInstr,
+        activation_type: ActivationType,
+        td_lookup: &HashMap<usize, onnx_tracer::trace_types::AtlasInstr>,
+        bytecode_preprocessing: &BytecodePreprocessing,
+    ) -> Self {
+        // Get input operand instruction
+        let input_instr = PreprocessingHelper::get_operand_instruction(
+            td_lookup,
+            instr.ts1,
+            &format!("{activation_type:?} operand"),
+        );
+
+        // Collect memory addresses
+        let input_addr = PreprocessingHelper::collect_and_pad(
+            input_instr,
+            bytecode_preprocessing,
+            &instr.output_dims,
+        );
+        let output_addr =
+            PreprocessingHelper::collect_and_pad(instr, bytecode_preprocessing, &instr.output_dims);
+
+        FpLookupInstance {
+            activation_type,
+            td_address: instr.address,
+            input_addr,
+            output_addr,
+            output_dims: instr.output_dims.clone(),
+        }
+    }
+}
+
+/// Compute the log2 of the LUT size needed to cover a given range
+/// The table covers signed integers from -2^(n-1) to 2^(n-1)-1
+pub fn compute_log_table_size(min_val: i32, max_val: i32) -> usize {
+    let abs_max = min_val.abs().max(max_val.abs()) as usize;
+    // We need n bits where 2^(n-1) > abs_max, i.e., n > log2(abs_max) + 1
+    let log_size = if abs_max == 0 {
+        1
+    } else {
+        (abs_max.next_power_of_two().trailing_zeros() as usize) + 1
+    };
+    // Clamp to maximum allowed size
+    log_size.min(MAX_LOG_FP_LOOKUP_TABLE_SIZE)
+}
+
+/// Generate a LUT tensor for the given activation type and table parameters.
+///
+/// This is a convenience function that uses the ActivationTable trait internally.
+/// For more control, use `get_activation_table(activation_type).materialize_tensor()`.
+pub fn generate_lut_tensor(
+    activation_type: ActivationType,
+    log_table_size: usize,
+    scale: f64,
+) -> Tensor<i32> {
+    get_activation_table(activation_type).materialize_tensor(log_table_size, scale)
+}
+
+/// Generate a multilinear polynomial from a LUT tensor.
+///
+/// This converts a Tensor<i32> (as returned by `generate_lut_tensor`) into a
+/// multilinear polynomial.
+pub fn generate_lut_polynomial<F: JoltField>(val_tensor: &Tensor<i32>) -> MultilinearPolynomial<F> {
+    let val: Vec<i64> = val_tensor
+        .data()
+        .to_vec()
+        .into_iter()
+        .map(|v| v as i64)
+        .collect();
+    MultilinearPolynomial::from(val)
+}
+
+/// Sumcheck instance for a single activation lookup.
+///
+/// Implements the shout protocol for small tables proving that activation outputs.
+/// Each instance handles one activation node.
+pub struct FpLookupSumcheck<F: JoltField> {
+    prover_state: Option<FpLookupProver<F>>,
+    input_claim: F,
+    num_rounds: usize,
+    tau: Vec<F>,
```

Comment:
Could we use an other name ? Such as `r_cycle`

```suggestion
    r_cycle: Vec<F>,
```